### PR TITLE
Employee Loan Management

### PR DESCRIPTION
**Pull request by KanchanChauhan**

### CODE REVIEW

File: erpnext/docs/user/manual/en/human-resources/employee-loan-management.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+ * Enter Employee and Loan Application
+ * Check "Include in Salary Slip" if the loan repayment will be deducted from the salary
+ * Enter Disbursement Date and Account Info
+ * As soon as you hit save, the EMI schedule is generated.
+ 
+<img class="screenshot" alt="EMI Schedule" src="{{docs_base_url}}/assets/img/human-resources/emi-schedule.png">
+
+#### Loan EMI deduction from Salary
+
```
Comment (rmehta) [inline erpnext/docs/user/manual/en/human-resources/employee-loan-management.md]: Don't use the term EMI - call it "Repayment". In general always avoid acronyms!

@KanchanChauhan @nabinhait

**Review by rmehta — CHANGES_REQUESTED** [review]

**Comment by KanchanChauhan** [comment]
@rmehta @nabinhait Done.

File: erpnext/hr/doctype/employee_loan/test_employee_loan.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+	
+	def	create_employee_loan(self):
+		if not frappe.db.get_value("Employee Loan", {"employee":self.employee}, "name"):
+			employee_loan = frappe.new_doc("Employee Loan")
+			employee_loan.update({
+					"employee": self.employee,
+					"company": erpnext.get_default_company(),
+					"posting_date": nowdate(),
+					"loan_type": "Personal Loan",
+					"rate_of_interest": 8.4,
```
Comment (nabinhait) [inline erpnext/hr/doctype/employee_loan/test_employee_loan.py]: Don't pass all these values, instead in `validate` of employee_loan, set those values if missing.

File: erpnext/hr/doctype/employee_loan/test_employee_loan.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+		employee_loan = frappe.get_doc("Employee Loan", {"employee":self.employee})
+		self.assertEquals(employee_loan.repayment_amount, 15052)
+		self.assertEquals(employee_loan.total_interest_payable, 21034)
+		self.assertEquals(employee_loan.total_payment, 301034)
+		self.assertEquals(len(employee_loan.repayment_schedule), 20)
+		self.assertEquals(employee_loan.repayment_schedule[3].principal_amount, 13369)
+		self.assertEquals(employee_loan.repayment_schedule[9].principal_amount, 13940)
+		self.assertEquals(employee_loan.repayment_schedule[19].principal_amount, 14941)
+		self.assertEquals(employee_loan.repayment_schedule[19].balance_loan_amount, 0)
+		self.assertEquals(employee_loan.repayment_schedule[14].balance_loan_amount, 73699)
```
Comment (nabinhait) [inline erpnext/hr/doctype/employee_loan/test_employee_loan.py]: Make it a for loop, for better readability.

File: erpnext/hr/doctype/employee_loan/test_employee_loan.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+		if not frappe.db.get_value("Employee Loan", {"employee":self.employee}, "name"):
+			employee_loan = frappe.new_doc("Employee Loan")
+			employee_loan.update({
+					"employee": self.employee,
+					"company": erpnext.get_default_company(),
+					"posting_date": nowdate(),
+					"loan_type": "Personal Loan",
+					"rate_of_interest": 8.4,
+					"loan_amount": 280000,
+					"repayment_method": "Repay Over Number of Periods",
```
Comment (nabinhait) [inline erpnext/hr/doctype/employee_loan/test_employee_loan.py]: Write a test case for "Repay Fixed Amount Every Month"

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 5]
```diff
@@ -329,11 +329,22 @@ def calculate_net_pay(self):
...
+		
+		self.set_loan_repayment()
 
-		self.net_pay = flt(self.gross_pay) - flt(self.total_deduction)
+		self.net_pay = flt(self.gross_pay) - (flt(self.total_deduction) + flt(self.loan_repayment))
 		self.rounded_total = rounded(self.net_pay,
 			self.precision("net_pay") if disable_rounded_total else 0)
 
+	def set_loan_repayment(self):
+		employee_loans = frappe.db.sql("""select * from `tabRepayment Schedule`
```
Comment (nabinhait) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: Never select *, use specific column names.

File: erpnext/hr/doctype/employee_loan/employee_loan.json [hunk 6]
```diff
@@ -0,0 +1,891 @@
...
+   "columns": 0, 
+   "fieldname": "include_in_salary_slip", 
+   "fieldtype": "Check", 
+   "hidden": 0, 
+   "ignore_user_permissions": 0, 
+   "ignore_xss_filter": 0, 
+   "in_filter": 0, 
+   "in_list_view": 0, 
+   "in_standard_filter": 0, 
+   "label": "Include in Salary Slip",
```
Comment (nabinhait) [inline erpnext/hr/doctype/employee_loan/employee_loan.json]: Repay from Salary?

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 7]
```diff
@@ -329,11 +329,22 @@ def calculate_net_pay(self):
...
+						where payment_date between %s and %s and parent in (select name from `tabEmployee Loan`
+						where employee = %s and include_in_salary_slip = 1 and docstatus = 1)""",
+						(self.start_date, self.end_date, self.employee), as_dict=True)
+		if employee_loans:
+			for employee_loan in employee_loans:
+				self.loan_repayment =+  employee_loan.total_payment
```
Comment (nabinhait) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: Why processing in python? Use `sum(repayment_amount)` in query.

**Review by nabinhait — CHANGES_REQUESTED** [review]
Accounting Entry missing on submission of Employee Loan and Salary Slip.

**Comment by nabinhait** [comment]
Test Case for loan amount in salary slip

**Review by rmehta — APPROVED** [review]

**Review by rmehta — APPROVED** [review]

**Pull request merged**